scalableminds/webknossos
# Refactor ping (prefetch) functionality

This PR refactors the ping (renamed to prefetch) functionality into a saga. Before, the prefetching was triggered after each rendering, whereas now, a standalone saga triggers the prefetching whenever the flycam was changed (position/rotation/zoom). The saga is throttled using the same throttle interval that was used to throttle the pinging before.
I've cleaned up other stuff while I was on it. One does get kind of "blind" during such a refactoring, because there's so much to fix, so if you have any other recommendations, please tell me and I'll incorporate them :)

Fun Performance/Memory Findings: During refactoring I noticed that there was a bug in the arbitary ping implementation, which caused prefetched buckets to always be in zoomstep 0. I fixed this and a simple test (20s of tracing in zoomstep > 0 in flight mode) showed that webknossos now downloads 20% less buckets :)
Also, the arbitraryPing used to be executed every single time the arbitrary view was rendered, even when no flycam property changed, which incurred quite a bit of processing overhead -> now a thing of the past :)

### ToDo:
  - [x] Add unit tests for prefetch_saga
  - Data is only prefetched after the first flycam action, should be triggered once initially - _Actually I don't think this is needed, once the user starts to move the prefetching starts (makes more sense anyways as we then know the movement direction)_

### URL of deployed dev instance (used for testing):
- `http://___.webknossos.xyz`

### Steps to test:
- Fly through the dataset, data should be prefetched and you should not need to wait for data to load during tracing

### Issues:
- part of #2601

------
